File: Codes/0040-combination-sum-ii.py
'''
Author: LetMeFly
Date: 2025-01-26 07:58:11
LastEditors: LetMeFly.xyz
LastEditTime: 2025-01-26 08:01:59
'''
from typing import List

class Solution:
    def dfs(self, target: int, index: int) -> None:
        if not target:
            self.ans.append([i for i in self.now])
            return
        if index == len(self.c) or target < 0:
            return
        self.now.append(self.c[index])
        self.dfs(target - self.c[index], index + 1)
        self.now.pop()
        next = index + 1
        while next < len(self.c) and self.c[next] == self.c[index]:
            next += 1
        self.dfs(target, next)
    
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        self.c = candidates
        self.ans = []
        self.now = []
        self.dfs(target, 0)
        return self.ans

File: Codes/0040-combination-sum-ii.java
/*
 * @Author: LetMeFly
 * @Date: 2025-01-26 08:02:41
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-01-26 08:10:08
 */
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    private List<List<Integer>> ans;
    private List<Integer> now;
    private int[] c;

    private void dfs(int target, int index) {
        if (target == 0) {
            ans.add(new ArrayList<>(now));
            return;
        }
        if (index == c.length || target < 0) {
            return;
        }
        now.add(c[index]);
        dfs(target - c[index], index + 1);
        now.remove(now.size() - 1);
        int next = index;
        while (++next < c.length && c[next] == c[index]);
        dfs(target, next);
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        c = candidates;
        ans = new ArrayList<>();
        now = new ArrayList<>();
        dfs(target, 0);
        return ans;
    }
}